### PR TITLE
No default qualifier stack trace output enhancement

### PR DESCRIPTION
**Pull request by jyluo**
In the scenarios where @DefaultQualifierInHierarchy or @DefaultFor(DefaultLocation.OTHERWISE) are not found in any of the qualifiers added to the supported qualifier set, or if that set is empty, the stack trace does not display how many or which qualifiers are examined. This fix updates the stack trace output to print out the number of qualifiers in the set as well as the canonical names of the qualifiers in an alphabetically sorted order. Below are some examples of the new stack trace output:

// No default qualifier:
com.sun.tools.javac.Main.main(Main.java:42)
  Underlying Exception: org.checkerframework.framework.source.SourceChecker$CheckerError: GenericAnnotatedTypeFactory.createQualifierDefaults: @DefaultQualifierInHierarchy or @DefaultFor(DefaultLocation.OTHERWISE) not found. Every checker must specify a default qualifier. 33 qualifiers examined: org.checkerframework.checker.units.qual.A, org.checkerframework.checker.units.qual.Acceleration, org.checkerframework.checker.units.qual.Angle, org.checkerframework.checker.units.qual.Area, org.checkerframework.checker.units.qual.C, org.checkerframework.checker.units.qual.Current, org.checkerframework.checker.units.qual.K, org.checkerframework.checker.units.qual.Length, org.checkerframework.checker.units.qual.Luminance, org.checkerframework.checker.units.qual.Mass, org.checkerframework.checker.units.qual.PolyUnit, org.checkerframework.checker.units.qual.Speed, org.checkerframework.checker.units.qual.Substance, org.checkerframework.checker.units.qual.Temperature, org.checkerframework.checker.units.qual.Time, org.checkerframework.checker.units.qual.UnitsBottom, org.checkerframework.checker.units.qual.UnknownUnits, org.checkerframework.checker.units.qual.cd, org.checkerframework.checker.units.qual.degrees, org.checkerframework.checker.units.qual.g, org.checkerframework.checker.units.qual.h, org.checkerframework.checker.units.qual.km2, org.checkerframework.checker.units.qual.kmPERh, org.checkerframework.checker.units.qual.m, org.checkerframework.checker.units.qual.m2, org.checkerframework.checker.units.qual.mPERs, org.checkerframework.checker.units.qual.mPERs2, org.checkerframework.checker.units.qual.min, org.checkerframework.checker.units.qual.mm2, org.checkerframework.checker.units.qual.mol, org.checkerframework.checker.units.qual.radians, org.checkerframework.checker.units.qual.s, org.checkerframework.framework.qual.PolyAll; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:667)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:28)

// No qualifiers at all in supportedQualifiers set:
  com.sun.tools.javac.Main.main(Main.java:42)
  Underlying Exception: org.checkerframework.framework.source.SourceChecker$CheckerError: GenericAnnotatedTypeFactory.createQualifierDefaults: @DefaultQualifierInHierarchy or @DefaultFor(DefaultLocation.OTHERWISE) not found. Every checker must specify a default qualifier. 0 qualifiers examined; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:667)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:28)

### CODE REVIEW

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 1]
```diff
@@ -754,17 +756,47 @@ public AnnotationFormatter getAnnotationFormatter() {
...
+     * supported by this checker. This method is useful for debug printing
+     * purposes, but otherwise returns the same set as
+     * {@link #getSupportedTypeQualifiers()}.
+     *
+     * @see #getSupportedTypeQualifiers()
+     *
+     * @return a sorted and immutable TreeSet of the supported type qualifiers,
+     *         or an empty set if no qualifiers are supported
+     */
+    public final Set<Class<? extends Annotation>> getSortedSupportedTypeQualifiers() {
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: It's weird that this method name is called getSORTED... when the return type is a Set, which by definition is not sorted.  Maybe it should return some sort of ordered collection rather than a Set.

Comment (jyluo) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: I'll change it to a List.

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 2]
```diff
@@ -754,17 +755,48 @@ public AnnotationFormatter getAnnotationFormatter() {
...
+     * @see #getSupportedTypeQualifiers()
+     *
+     * @return an immutable sorted list of the supported type qualifiers, or an
+     *         immutable empty list if no qualifiers are supported
+     */
+    public final List<Class<? extends Annotation>> getSortedSupportedTypeQualifiers() {
+        // insert into an array list then sort it alphabetically
+        List<Class<? extends Annotation>> sortedSupportedQuals = new ArrayList<Class<? extends Annotation>>();
+        sortedSupportedQuals.addAll(getSupportedTypeQualifiers());
+        sortedSupportedQuals.sort(QUALIFIER_SORT_ORDERING);
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: List does not have a `sort()` method, so this code does not compile.

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 3]
```diff
@@ -354,6 +354,39 @@ protected final QualifierDefaults createQualifierDefaults() {
...
+     * @return a string containing the number of qualifiers and canonical names
+     *         of each qualifier.
+     */
+    private final String getSortedQualifierNames() {
+        List<Class<? extends Annotation>> sortedQuals = getSortedSupportedTypeQualifiers();
+
+        // display the number of qualifiers as well as the names of each
+        // qualifier.
+        StringBuilder sb = new StringBuilder();
+        sb.append(" ");
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: I would not start the string with a space or mention somewhere that it does.

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 4]
```diff
@@ -754,17 +755,48 @@ public AnnotationFormatter getAnnotationFormatter() {
...
+     * supported by this checker. This method is useful for debug printing
+     * purposes, but otherwise returns the same qualifiers as
+     * {@link #getSupportedTypeQualifiers()}.
+     *
+     * @see #getSupportedTypeQualifiers()
+     *
+     * @return an immutable sorted list of the supported type qualifiers, or an
+     *         immutable empty list if no qualifiers are supported
+     */
+    public final List<Class<? extends Annotation>> getSortedSupportedTypeQualifiers() {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: I don't see the value of having this public method in the ATF. I would fold the logic into getSortedQualifierNames or have a private method there.

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 5]
```diff
@@ -354,6 +354,39 @@ protected final QualifierDefaults createQualifierDefaults() {
...
 
     /**
+     * Creates and returns a string containing the number of qualifiers and the
+     * canonical class names of each qualifier that has been added to this
+     * checker's supported qualifier set. The names are alphabetically sorted.
+     *
+     * @return a string containing the number of qualifiers and canonical names
+     *         of each qualifier.
+     */
+    private final String getSortedQualifierNames() {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: Why is this method private and not protected? Might errors in subclasses use this method?

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 6]
```diff
@@ -754,17 +755,48 @@ public AnnotationFormatter getAnnotationFormatter() {
...
-     * {@link #createSupportedTypeQualifiers createSupportedTypeQualifiers} instead.
-
+     * {@link #createSupportedTypeQualifiers createSupportedTypeQualifiers}
+     * instead.
+     *
      * @see #createSupportedTypeQualifiers()
      *
-     * @return the type qualifiers supported this processor, or an empty
-     * set if none
+     * @return an immutable HashSet of the supported type qualifiers, or an
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: Why the change to HashSet? It's a set and no ordering or concrete implementation is prescribed.

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java (hunk 3 above)
Comment (jyluo) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: I'll remove the space here.
